hpe3parclient/client.py

```diff
@@ -508,46 +508,65 @@ def createVolume(self, name, cpgName, sizeMiB, optional=None):
+                        Deco volumes, both 'tdvv' and 'compression' \
+                        must be specified")
+                if optional.get('tdvv') is True \
                         and optional.get('compression') is True:
                     optional['reduce'] = True
 
-                if not optional.get('tpvv') \
-                        and not optional.get('compression'):
-                    optional['tpvv'] = True
+                if optional.get('tdvv') is False \
+                        and optional.get('compression') is False:
+                    optional['reduce'] = False
 
-                if 'compression' in optional:
-                    if optional.get('compression') is not None:
-                        if optional.get('compression') is True:
-                            optional['reduce'] = True
-                        elif optional.get('compression') is False:
-                            optional['reduce'] = False
-                        else:
-                            # raising exception for junk compression input
-                            ex_desc = "39 - invalid input: wrong type for value \
-                                       - compression"
-                            raise exceptions.HTTPBadRequest(ex_desc)
+                if optional.get('tdvv') is False \
+                        and optional.get('compression') is True:
+                    raise exceptions.HTTPBadRequest("For enabling the \
+                        compression for primera tdvv and compression \
+                        both flags must be true")
 
-                    optional.pop('compression')
+                if optional.get('tdvv') is True \
+                        and optional.get('compression') is False:
+                    raise exceptions.HTTPBadRequest("For enabling the \
+                        compression for primera tdvv and compression \
+                        both are true")
 
-                if optional.get('tpvv') is True \
-                        and optional.get('reduce') is True:
-                    optional.pop('tpvv')
+                if 'compression' in optional:
+                    optional.pop('compression')
+                if 'tdvv' in optional:
+                    optional.pop('tdvv')
             info = self._mergeDict(info, optional)
         logger.debug("Parameters passed for create volume %s" % info)
-
-        response, body = self.http.post('/volumes', body=info)
-        return body
+        try:
+            response, body = self.http.post('/volumes', body=info)
+            return body
+        except Exception as ex:
+            if self.primera_supported:
+                ex_desc = ex.get_description()
+                if re.search("invalid input: one of the parameters is \
+                        required - tpvv,reduce", ex_desc):
+                    ex_desc = "invalid input: one of the parameters is \
+                        required - tpvv,compression"
```

Comment:
Invalid input: either tpvv must be True OR tdvv and compression must be True. Both cannot be False.